pingcap/tidb
# config, sysvar: map `max-server-connections` to `max_connections`

### What problem does this PR solve?


Issue Number: ref #34960

Problem Summary:

Before, if we want to change the limitation of `max-server-connections`, we have to change the configuration and restart the node. And the system variable `max_connections` is not available.

### What is changed and how it works?

Add a new config item `instance.max_connections` and make sysvar `max_connections` available.

### Check List

Tests 

- [x] Unit test (using previous testing)
> Thanks @morgo for suggesting this manual test!
```
Test 1: check whether 'max_connections' limits the number of incoming connections:
1. start a tidb-server
2. start one mysql client connecting to the server with unlimited 'max_connections'
3. 'set @@global.max_connections = 3'
4. try to start other mysql clients connecting to the server one by one, until some errors happen.
Result: only three clients can connect to the server.
The later connections encounter 'ERROR 1040 (08004): Too many connections'

Test 2: check whether 'set @@global.max_connections = ?' success:
1. start a tidb-server
2. start 5 mysql clients connecting to the server with unlimited 'max_connections'
3. try to 'set @@global.max_connections = ?' using the number from 0 to 10
Result: whatever the value is, the setting is always successful. 
But if the setting value is bigger than the current connection number, no other connections can come in.
This behaviour is the same with MySQL.
```

Side effects


Documentation

> For [MySQL 8.0](https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_max_connections), the range of `max_connections` is `[1, 100000]`, and the default value is `151`. In TiDB, since the default value of `instance.max_connections` is `0`, which means *no limitation*, so the range of `max_connections` in TiDB is `[0, 100000]`, and default value is `0`.

### Release note



Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```
